### ISSUE TITLE
Disabled plugins get picked up by PackerClean

### ISSUE THREAD

**Issue by nanotee** [wbthomason/packer.nvim#56]
I sometimes want to disable plugins along with their associated configuration while still keeping them installed and up to date. I assumed that was the goal of the `disable` option but the `:PackerClean` command asks me if I want to remove plugins that are disabled. Am I misunderstanding the purpose of `disable`?

**Comment by wbthomason** [comment]
Ah, I should clarify this in the docs. Yes, `disable` is meant as a mechanism for temporarily removing a plugin - it is by design that `PackerClean` removes disabled plugins. For keeping a plugin around and up to date, but not loading it, it is intended to use `opt = true` (manually setting the plugin as `opt`).

I can see how this would be confusing, though. Perhaps I should make `disable` effectively just set `opt = true` and add some other keyword (though I don't know what would be a good choice for this) for the current `disable` functionality.

**Comment by clason** [comment]
Note that you can't rely on `opt=true` disabling the plugin, since there might be a `packadd` somewhere else in the config...

The safest way to handle that would be to move disabled plugins to another `pack/**/disabled` directory.

**Comment by wbthomason** [comment]
That's true - my intent was that if a user manually marks `opt = true`, then they are responsible for loading or not loading it manually, as they prefer. Perhaps that's also a bad assumption.

**Comment by clason** [comment]
That is not an unreasonable assumption. (In my case, it's because they're lua plugins so I need to manually load them anyway to configure them from init.vim.)

Regarding `disable = true`, what does this differently from just commenting out the `use` line in the config? (I guess that is the source of the confusion -- people would expect it to do something different since commenting out is easier than adding the keyword?)

**Comment by wbthomason** [comment]
That's fair. The difference is mostly in a capability I don't think I've ever explicitly described: `disable` enables conditional installation. Consider the following:

Imagine you use both Windows and Linux machines, and there's some plugin `foobar/WindowsVim` that you only like to use on Windows. You could then write:
```lua
use { 'foobar/WindowsVim', disable = test_for_windows() }
```
and use your configuration across both of your machines without needing to edit it, while still only having `WindowsVim` be installed on Windows.

There are, of course, other ways to do this - you could install but not load `WindowsVim` by using the `cond` keyword, or you could just wrap the `use` call in a conditional, e.g.
```lua
if test_for_windows() then
  use 'foobar/WindowsVim'
end
```
but I find `disable` to be more concise and a little nicer.

**Comment by wbthomason** [comment]
Regardless, it's clear the current name for that feature is confusing, and there's a confusing overlap in functionality between `disable`, `cond`, and `opt = true`. I'm very open to suggestions for improvement; I'll also think about how to clarify this.

**Comment by clason** [comment]
Thanks, that makes sense, and I agree that the conditional installation (as opposed to configuration/activation) via `disable` feels more "lua-y" than the external conditional (which people would be familiar with from `vim-plug`).

But I think you put your finger on it with the overlap. I think one helpful first step is to add a brief explanation -- like you have for `dependencies` --  on what `disable` actually does (i.e., treat it as if the `use` entry was not present, in particular leading to removal on `clean`, differently from `cond)  and that the right-hand side can actually be a lua function that returns a boolean. Especially the latter makes it clear that it is strictly more powerful than simply uncommenting -- and I certainly didn't realize that this was possible ;)

**Comment by wbthomason** [comment]
> Another point leading to some mental friction is having an option that disables something you've just written; I think it would feel more natural to turn this around to have an enable option (that defaults to true) -- this would also be more consistent with the cond option. It would also help to have the names be more similar -- enable and load, for example, or cond_enable and cond_load. (I'd prefer the former.)

Yes, I was thinking this too - `enable` and `load` make much more sense than `disable` and `cond`.

> Finally, the difference between "loading" and "enabling" is not clear from the documentation -- the latter I now understand after your comment, but what exactly does the former mean?

By "loading" I typically mean "sourcing in Neovim", whereas "enabling" typically means "installing". I have probably used each interchangeably/incorrectly per my definitions here in the docs; I'll do a consistency pass to correct that.

(to further clarify: "loading" is what happens when e.g. an opt plugin is `packadd`ed, or a start plugin is sourced during Neovim startup)

**Comment by nanotee** [comment]
Ah, I see why I was confused now, thank you for the clarification.

For context, I was trying to do something like this:

```lua
use {
    'user/plugin',
    cond = 'not vim.g.minimal_config',
}
```

Using `opt = true` in this case doesn't prevent the plugin from loading entirely because the condition automatically makes the plugin `opt` and *will* load the plugin if it is true.

What I wanted basically amounts to "don't load the plugin even if the condition is met and keep it up to date", but I suppose I could just as well comment the condition out and add `opt = true`.

I don't know how many users would actually do something like that so perhaps my use case is a bit niche :)

**Comment by wbthomason** [comment]
Ok, I understand. That is a use case I admit I had not considered.

There is another underdocumented feature that might help you: adding your own handlers.

By passing a name and a function with signature (table of plugins, plugin table, arbitrary value specific to your handler) to `packer.set_handler`, you can add a "keyword" to `packer`. You could imagine here adding something like the following for your use case (note: I have not tested this; it may well not work):
```lua
packer.set_handler('never_load', function(plugins, plugin, value)
  if value then
    local compile = require('packer.compile')
    -- Remove anything that could lead to the plugin getting lazy-loaded
    for _, k in ipairs(compile.opt_keys) do
      plugin[k] = nil
    end
    plugin.config = nil
    -- Ensure the plugin isn't eager-loaded
    plugin.opt = true
    plugin.manual_opt = true -- This is only necessary due to a dumb hack I need to eventually remove 
  end
end)
```
and then writing:
```lua
use {
    'user/plugin',
    cond = 'not vim.g.minimal_config',
    never_load = true
}
```

**Comment by nanotee** [comment]
I haven't had the chance to play with that feature yet, I may be able to contribute some documentation if I can come up with useful examples. Thank you for the suggestion

**Comment by akinsho** [comment]
Hey, so I just found this issue and have exactly the use case @wbthomason described in the usage of `disable` i.e. some plugins I want to specify but only have installed/loaded on some machines, I thought maybe I could help document the usage, but wasn't sure if the API might change based on this discussion?.

Secondly I noted that packer tries to clean the disabled plugins which is fine given the described use case but it doesn't actually seem to remove them so every call to `PackerClean` with disabled plugins seems to just list them again. Not sure if this is a recent regression, long standing bug or something about the functionality that I've missed?
